Feat: Internal staff approve/reject Operation request for carbon tax exemption

andrea-williams opened the pull request:
Feat: Implemented Approve/Reject operation functionality with unit tests

[ZH card 97](https://app.zenhub.com/workspaces/team-moose-sprint-board-60d216175a2689000ec774d0/issues/gh/bcgov/cas-registration/97)

![Screenshot 2023-11-01 at 11 50 43 AM](https://github.com/bcgov/cas-registration/assets/54914403/5c83cc3b-9754-44ba-9820-3351cd64e43b)

![Screenshot 2023-11-01 at 11 51 34 AM](https://github.com/bcgov/cas-registration/assets/54914403/32ed78ef-719a-4d35-a8d4-f15e82bc308d)

shon-button commented:
Nit: Status displays as lower case rather than proper case (approved; rejected) in [Operations](http://localhost:3000/dashboard/operations)

Q: Should the user be able to Submit the form?
If no, then the "Submit" should not display.
If yes, then perhaps the Review component should not be displayed with the confirm message?
![image](https://github.com/bcgov/cas-registration/assets/120038448/86a0e8ba-22a4-4ff3-ae10-5ec5a9da5a44)

Q: After Accept\Reject action, should the buttons display on subsequent "View Details" 
If yes, should the Accept\Reject button's disable status reflect the Accept\Reject status
Q: Should the the Accept\Reject status display on subsequent "View Details"

JoshLarouche commented on bc_obps/registration/api.py:
```diff
@@ -166,12 +191,18 @@ def request_access(request, payload: SelectOperatorIn):
 
 
-@router.get("/select-operator/user-operator/{int:user_operator_id}", response=UserOperatorOut)
+@router.get(
```
It looks like there was a rebase error here that changed the structure of the endpoint

JoshLarouche commented on bc_obps/registration/api.py:
```diff
@@ -217,7 +248,7 @@ def get_user_operator(request, user_operator_id: int):
 
 @router.put(
-    "/select-operator/user-operator/{int:user_operator_id}",
+    "/select-operator/request-access/user-operator/{int:user_operator_id}",
```
Same rebase error here

JoshLarouche replied:
Looks like this rebase error is still here

JoshLarouche commented on bc_obps/registration/tests/test_endpoints.py:
```diff
@@ -81,70 +88,157 @@ def test_post_new_operation(self, client):
-            name='New name',
-            type='Single Facility Operation',
+            name="New name",
+            type="Type 1",
```
rebase error here I think

andrea-williams commented:
Updates: 

- fixed the rebasing errors that @JoshLarouche  caught (thank you!)
- fixed the bug that @shon-button noticed (great catch!) - the Review component won't appear when creating a new Operation - now it only appears when the Operation's status is Pending (although there maaaaayyyy be a need to change this later, for example, if an internal user accidentally rejects an Operation they intended to Approve, there's no way to undo that)
- the lowercase 'accepted' or 'rejected' status comes from our database. I've added a function to `Operations` component so that the first letter of the status appears capitalized in the table

JoshLarouche commented on .github/workflows/test.yaml:
```diff
@@ -233,6 +233,8 @@ jobs:
         with:
           django_secret_key: ${{ env.DJANGO_SECRET_KEY }}
+      - name: Setup tmate session
+        uses: mxschmitt/action-tmate@v3
```
I think this needs to be removed

JoshLarouche commented on bc_obps/registration/api/user_operator.py:
```diff
@@ -63,7 +66,7 @@ def get_user_operator(request, user_operator_id: int):
 
 @router.put(
-    "/select-operator/user-operator/{int:user_operator_id}",
+    "/select-operator/request-access/user-operator/{int:user_operator_id}",
```
I think there still might be a rebasing issue here it looks like the route was changed

dleard commented:
Have we implemented any accessibility checks in our e2e tests? I ask because red/green buttons can be troublesome to get the ratios right to accomodate the most common colour blindness

dleard commented on bc_obps/bc_obps/settings.py:
```diff
@@ -26,7 +26,7 @@
 
 # SECURITY WARNING: keep the secret key used in production secret!
-SECRET_KEY = os.environ.get("DJANGO_SECRET_KEY")
+SECRET_KEY = os.environ.get("SECRET_KEY")
```
did this change again? It was 'SECRET_KEY', and then a PR last week changed it to 'DJANGO_SECRET_KEY` https://github.com/bcgov/cas-registration/pull/234

andrea-williams replied:
this is why I need rebasing coaching! No, it didn't change again. Git is just playing monkey business and I'm too incompetent to catch them at it

JoshLarouche commented on client/app/components/routes/operations/form/Operation.tsx:
```diff
@@ -4,6 +4,13 @@ import OperationsForm, {
+=======
+// 📚 runtime mode for dynamic data to allow build w/o api
+export const runtime = "edge";
+>>>>>>> 1c86db3 (Feat: Implemented Approve/Reject operation functionality with unit tests; 2 other unrelated tests failing)
```
looks like merge conflict lines got added in here

dleard commented on bc_obps/registration/api/operation.py:
```diff
@@ -20,6 +24,18 @@ def get_operation(request, operation_id: int):
+    operation = get_object_or_404(Operation, id=operation_id)
+    return operation
+
+
```
Is this all duplicate code? Looks like 3 implementations of the same endpoint?

dleard commented on client/app/components/routes/operations/form/Operation.tsx:
```diff
@@ -4,6 +4,13 @@ import OperationsForm, {
 import { RJSFSchema } from "@rjsf/utils";
 import { fetchAPI } from "@/app/utils/api";
+import Review from "./Review";
+<<<<<<< HEAD
```
leftover rebase syntax

andrea-williams commented:
> Have we implemented any accessibility checks in our e2e tests? I ask because red/green buttons can be troublesome to get the ratios right to accomodate the most common colour blindness

at the moment, no. We've got a ticket in our backlog (that will probably end up being split into multiple tickets) to address accessibility accomodations. Right now we're just focused on getting functional features in. Polishing the UI has been de-prioritized for the moment, although we can start thinking about it now that we've got Marcel.

dleard commented:
> > Have we implemented any accessibility checks in our e2e tests? I ask because red/green buttons can be troublesome to get the ratios right to accomodate the most common colour blindness
> 
> at the moment, no. We've got a ticket in our backlog (that will probably end up being split into multiple tickets) to address accessibility accomodations. Right now we're just focused on getting functional features in. Polishing the UI has been de-prioritized for the moment, although we can start thinking about it now that we've got Marcel.

The 'thumbs up' icon and text (approve/deny) do make it clear what the actions are which mitigates the accessibility issue. Just wanted to flag the green/red buttons as something we may want to revisit. For consistency in CIF, we used a theme throughout the app where all buttons were either solid blue (primary) or outline blue (secondary)

JoshLarouche reviewed (approved):
All rebasing leftovers seem to be gone good job! One final nitpick, it looks like your success and reject response messages disappear so fast they aren't really showing the user feedback. Having them disappear after a delay or stay on the page until refresh (I think it may have worked like this before but not sure) would be better.
